firewood/src/db.rs

```diff
@@ -305,13 +306,64 @@ impl<S: ShaleStore<Node> + Send + Sync> api::DbView for DbRev<S> {
             .map_err(|e| api::Error::IO(std::io::Error::new(ErrorKind::Other, e)))
     }
 
-    async fn range_proof<K: api::KeyType, V, N>(
+    async fn range_proof<K: api::KeyType, V>(
         &self,
-        _first_key: Option<K>,
-        _last_key: Option<K>,
-        _limit: usize,
-    ) -> Result<Option<api::RangeProof<K, V, N>>, api::Error> {
-        todo!()
+        first_key: Option<K>,
+        last_key: Option<K>,
+        limit: Option<usize>,
+    ) -> Result<Option<api::RangeProof<Vec<u8>, Vec<u8>>>, api::Error> {
+        let mut stream = self.stream(first_key.as_ref())?;
+
+        // fetch the first key from the stream
+        let first_result = stream.next().await;
+
+        // transpose the Option<Result<T, E>> to Result<Option<T>, E>
+        // If this is an error, the ? operator will return it
+        let Some((key, _)) = first_result.transpose()? else {
+            // nothing returned, either the trie is empty or the key wasn't found
+            return Ok(None);
+        };
+
+        let Some(first_key) = self.single_key_proof(key).await? else {
+            return Ok(None);
+        };
+
+        // we stop streaming if either we hit the limit or the key returned was larger
+        // than the largest key requested
+        let mut middle = stream
+            .take(limit.unwrap_or(usize::MAX))
```

Comment:
This needs to come after the `take_while`

Reply:
See my comment above about `merkle.rs`. If we put this functionality there, it's also easy to test. we should have tests before merging. Not necessarily tests for proving a range-proof, but tests that give us confidence that the filters work properly (first-key, last-key, and limit).

Reply:
Ah, you're right, my bad. I was thinking of `filter` (my brain was playing tricks on me).